### PR TITLE
POC delete stale non-e2e devices for users

### PR DESCRIPTION
**Pull request by erikjohnston**
This should help reduce the number of devices e.g. simple bots the repeatedly login rack up. 

We only delete non-e2e devices as they should be safe to delete, whereas if we delete e2e devices for a user we may accidentally break their ability to receive e2e keys for a message.

### CODE REVIEW

File: synapse/storage/databases/main/devices.py [hunk 1]
```diff
@@ -1466,6 +1466,40 @@ def _txn(txn: LoggingTransaction) -> int:
...
+
+        num_devices = await self.db_pool.simple_select_one_onecol(
+            table="devices",
+            keyvalues={"user_id": user_id, "hidden": "false"},
+            retcol="COALESCE(COUNT(*), 0)",
+            desc="count_devices",
+        )
+
+        if num_devices > 10:
+            return ()
```
Comment (anoadragon453) [inline synapse/storage/databases/main/devices.py:1480]: Did you mean < 10?

File: synapse/storage/databases/main/devices.py [hunk 2]
```diff
@@ -1466,6 +1466,40 @@ def _txn(txn: LoggingTransaction) -> int:
 
         return rows
 
+    async def get_stale_devices(self, user_id: str) -> Collection[str]:
+        """Get set of devices that we could delete for this user."""
+
+        num_devices = await self.db_pool.simple_select_one_onecol(
+            table="devices",
+            keyvalues={"user_id": user_id, "hidden": "false"},
```
Comment (clokep) [inline synapse/storage/databases/main/devices.py:1474]:
```suggestion
            keyvalues={"user_id": user_id, "hidden": False},
```

Comment (erikjohnston) [reply synapse/storage/databases/main/devices.py]: :facepalm:

**Comment by richvdh** [comment]
I believe this is intended as a mitigation to https://github.com/matrix-org/synapse/issues/8263.

**Comment by richvdh** [comment]
Isn't this the same as https://github.com/matrix-org/synapse/pull/12855? Or at least, closely related?

File: synapse/handlers/device.py [hunk 3]
```diff
@@ -433,6 +435,14 @@ async def check_device_registered(
...
 
+    async def _delete_stale_devices(self, user_id: str) -> None:
+        """Delete any stale devices this user may have."""
+        device_ids = await self.store.get_stale_devices(user_id)
+        if not device_ids:
+            return
+
+        await self.delete_devices(user_id, device_ids)
+
     async def _delete_stale_devices(self) -> None:
```
Comment (richvdh) [inline synapse/handlers/device.py:446]: erm, two methods both called `_delete_stale_devices` ?

**Comment by richvdh** [comment]
> Isn't this the same as #12855? Or at least, closely related?

looks like they are taking slightly different approaches, but I worry that having both approaches in the codebase is going to be confusing for everyone.

**Comment by erikjohnston** [comment]
I've updated this PR to change the names of the functions, to hopefully make it much less confusing w.r.t. the existing deleting stale devices option.

I've also made it such that we have a hard limit of 50 non-e2e devices, as two weeks worth of devices can still be a lot.

File: synapse/storage/databases/main/devices.py [hunk 4]
```diff
@@ -1526,6 +1526,70 @@ def _txn(txn: LoggingTransaction) -> int:
...
+
+        num_devices = await self.db_pool.simple_select_one_onecol(
+            table="devices",
+            keyvalues={"user_id": user_id, "hidden": False},
+            retcol="COALESCE(COUNT(*), 0)",
+            desc="count_devices",
+        )
+
+        # We let users have up to ten devices without pruning.
+        if num_devices < 10:
```
Comment (squahtx) [inline synapse/storage/databases/main/devices.py:1544]: I think this was the intention?
```suggestion
        # We let users have up to ten devices without pruning.
        if num_devices <= 10:
```

**Review by squahtx — APPROVED** [review]

File: synapse/handlers/device.py [hunk 5]
```diff
@@ -452,6 +455,14 @@ async def check_device_registered(
 
         raise errors.StoreError(500, "Couldn't generate a device ID.")
 
+    async def _prune_too_many_devices(self, user_id: str) -> None:
+        """Delete any stale devices this user may have."""
```
Comment (squahtx) [inline synapse/handlers/device.py:459]: Echoing [Rich's comment](https://github.com/matrix-org/synapse/pull/14038#issuecomment-1274618835).
It's mildly confusing that we have two concepts of stale devices in the codebase. Let's avoid using the same term to refer to both.

```suggestion
        """Delete any excess old devices this user may have."""
```

**Comment by richvdh** [comment]
If this is a POC... should it really have been merged and deployed to matrix.org?

**Comment by erikjohnston** [comment]
We should have updated the summary after more work was done.

**Comment by richvdh** [comment]
I believe this has been revered (by #14582)

**Pull request merged**